// @rsksmart/rsk3/dist/@rsksmart/rsk3.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('@babel/runtime/helpers/classCallCheck'), require('@babel/runtime/helpers/createClass'), require('@babel/runtime/helpers/set'), require('@babel/runtime/helpers/get'), require('@babel/runtime/helpers/possibleConstructorReturn'), require('@babel/runtime/helpers/getPrototypeOf'), require('@babel/runtime/helpers/inherits'), require('web3-core'), require('web3-providers'), require('web3-core-helpers'), require('@rsksmart/rsk3-account'), require('@rsksmart/rsk3-contract'), require('@rsksmart/rsk3-personal'), require('@rsksmart/rsk3-abi'), require('@rsksmart/rsk3-net'), require('@rsksmart/rsk3-utils'), require('@babel/runtime/regenerator'), require('@babel/runtime/helpers/asyncToGenerator'), require('ethereumjs-tx'), require('lodash/isString'), require('web3-core-method'), require('web3-core-subscriptions'), require('lodash')) :
    typeof define === 'function' && define.amd ? define(['@babel/runtime/helpers/classCallCheck', '@babel/runtime/helpers/createClass', '@babel/runtime/helpers/set', '@babel/runtime/helpers/get', '@babel/runtime/helpers/possibleConstructorReturn', '@babel/runtime/helpers/getPrototypeOf', '@babel/runtime/helpers/inherits', 'web3-core', 'web3-providers', 'web3-core-helpers', '@rsksmart/rsk3-account', '@rsksmart/rsk3-contract', '@rsksmart/rsk3-personal', '@rsksmart/rsk3-abi', '@rsksmart/rsk3-net', '@rsksmart/rsk3-utils', '@babel/runtime/regenerator', '@babel/runtime/helpers/asyncToGenerator', 'ethereumjs-tx', 'lodash/isString', 'web3-core-method', 'web3-core-subscriptions', 'lodash'], factory) :
    (global = global || self, global.Rsk3 = factory(global._classCallCheck, global._createClass, global._set, global._get, global._possibleConstructorReturn, global._getPrototypeOf, global._inherits, global.web3Core, global.web3Providers, global['web3-core-helpers'], global.rsk3Account, global.rsk3Contract, global.rsk3Personal, global.rsk3Abi, global.rsk3Net, global.Utils, global._regeneratorRuntime, global._asyncToGenerator, global.EthereumTx, global.isString, global.web3CoreMethod, global.web3CoreSubscriptions, global._));
}(this, (function (_classCallCheck, _createClass, _set, _get, _possibleConstructorReturn, _getPrototypeOf, _inherits, web3Core, web3Providers, web3CoreHelpers, rsk3Account, rsk3Contract, rsk3Personal, rsk3Abi, rsk3Net, Utils, _regeneratorRuntime, _asyncToGenerator, EthereumTx, isString, web3CoreMethod, web3CoreSubscriptions, _) { 'use strict';

    _classCallCheck = _classCallCheck && Object.prototype.hasOwnProperty.call(_classCallCheck, 'default') ? _classCallCheck['default'] : _classCallCheck;
    _createClass = _createClass && Object.prototype.hasOwnProperty.call(_createClass, 'default') ? _createClass['default'] : _createClass;
    _set = _set && Object.prototype.hasOwnProperty.call(_set, 'default') ? _set['default'] : _set;
    _get = _get && Object.prototype.hasOwnProperty.call(_get, 'default') ? _get['default'] : _get;
    _possibleConstructorReturn = _possibleConstructorReturn && Object.prototype.hasOwnProperty.call(_possibleConstructorReturn, 'default') ? _possibleConstructorReturn['default'] : _possibleConstructorReturn;
    _getPrototypeOf = _getPrototypeOf && Object.prototype.hasOwnProperty.call(_getPrototypeOf, 'default') ? _getPrototypeOf['default'] : _getPrototypeOf;
    _inherits = _inherits && Object.prototype.hasOwnProperty.call(_inherits, 'default') ? _inherits['default'] : _inherits;
    _regeneratorRuntime = _regeneratorRuntime && Object.prototype.hasOwnProperty.call(_regeneratorRuntime, 'default') ? _regeneratorRuntime['default'] : _regeneratorRuntime;
    _asyncToGenerator = _asyncToGenerator && Object.prototype.hasOwnProperty.call(_asyncToGenerator, 'default') ? _asyncToGenerator['default'] : _asyncToGenerator;
    EthereumTx = EthereumTx && Object.prototype.hasOwnProperty.call(EthereumTx, 'default') ? EthereumTx['default'] : EthereumTx;
    isString = isString && Object.prototype.hasOwnProperty.call(isString, 'default') ? isString['default'] : isString;
    _ = _ && Object.prototype.hasOwnProperty.call(_, 'default') ? _['default'] : _;

    var TransactionSigner = function () {
      function TransactionSigner(utils, formatters) {
        _classCallCheck(this, TransactionSigner);
        this.utils = utils;
        this.formatters = formatters;
      }
      _createClass(TransactionSigner, [{
        key: "sign",
        value: function () {
          var _sign = _asyncToGenerator( _regeneratorRuntime.mark(function _callee(transaction, privateKey) {
            var ethTx, validationResult, rlpEncoded, rawTransaction, transactionHash;
            return _regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    if (privateKey) {
                      _context.next = 2;
                      break;
                    }
                    throw new Error('No privateKey given to the TransactionSigner.');
                  case 2:
                    if (isString(privateKey)) {
                      _context.next = 4;
                      break;
                    }
                    throw new Error('privateKey should be string type');
                  case 4:
                    if (privateKey.startsWith('0x')) {
                      privateKey = privateKey.substring(2);
                    }
                    ethTx = new EthereumTx(transaction);
                    ethTx.sign(Buffer.from(privateKey, 'hex'));
                    validationResult = ethTx.validate(true);
                    if (!(validationResult !== '')) {
                      _context.next = 10;
                      break;
                    }
                    throw new Error("TransactionSigner Error: ".concat(validationResult));
                  case 10:
                    rlpEncoded = ethTx.serialize().toString('hex');
                    rawTransaction = '0x' + rlpEncoded;
                    transactionHash = this.utils.keccak256(rawTransaction);
                    return _context.abrupt("return", {
                      messageHash: Buffer.from(ethTx.hash(false)).toString('hex'),
                      v: '0x' + Buffer.from(ethTx.v).toString('hex'),
                      r: '0x' + Buffer.from(ethTx.r).toString('hex'),
                      s: '0x' + Buffer.from(ethTx.s).toString('hex'),
                      rawTransaction: rawTransaction,
                      transactionHash: transactionHash
                    });
                  case 14:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee, this);
          }));
          function sign(_x, _x2) {
            return _sign.apply(this, arguments);
          }
          return sign;
        }()
      }, {
        key: "type",
        get: function get() {
          return 'TransactionSigner';
        }
      }]);
      return TransactionSigner;
    }();

    function _createSuper(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var GetBlockMethod = function (_AbstractGetBlockMeth) {
      _inherits(GetBlockMethod, _AbstractGetBlockMeth);
      var _super = _createSuper(GetBlockMethod);
      function GetBlockMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, GetBlockMethod);
        return _super.call(this, 'eth_getBlockByNumber', utils, formatters, moduleInstance);
      }
      _createClass(GetBlockMethod, [{
        key: "beforeExecution",
        value: function beforeExecution(moduleInstance) {
          if (this.isHash(this.parameters[0])) {
            this.rpcMethod = 'eth_getBlockByHash';
          }
          _get(_getPrototypeOf(GetBlockMethod.prototype), "beforeExecution", this).call(this, moduleInstance);
        }
      }]);
      return GetBlockMethod;
    }(web3CoreMethod.AbstractGetBlockMethod);

    function _createSuper$1(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$1()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$1() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var GetUncleMethod = function (_AbstractGetUncleMeth) {
      _inherits(GetUncleMethod, _AbstractGetUncleMeth);
      var _super = _createSuper$1(GetUncleMethod);
      function GetUncleMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, GetUncleMethod);
        return _super.call(this, 'eth_getUncleByBlockNumberAndIndex', utils, formatters, moduleInstance);
      }
      _createClass(GetUncleMethod, [{
        key: "beforeExecution",
        value: function beforeExecution(moduleInstance) {
          if (this.isHash(this.parameters[0])) {
            this.rpcMethod = 'eth_getUncleByBlockHashAndIndex';
          }
          _get(_getPrototypeOf(GetUncleMethod.prototype), "beforeExecution", this).call(this, moduleInstance);
        }
      }]);
      return GetUncleMethod;
    }(web3CoreMethod.AbstractGetUncleMethod);

    function _createSuper$2(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$2()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$2() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var GetBlockTransactionCountMethod = function (_AbstractGetBlockTran) {
      _inherits(GetBlockTransactionCountMethod, _AbstractGetBlockTran);
      var _super = _createSuper$2(GetBlockTransactionCountMethod);
      function GetBlockTransactionCountMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, GetBlockTransactionCountMethod);
        return _super.call(this, 'eth_getBlockTransactionCountByNumber', utils, formatters, moduleInstance);
      }
      _createClass(GetBlockTransactionCountMethod, [{
        key: "beforeExecution",
        value: function beforeExecution(moduleInstance) {
          if (this.isHash(this.parameters[0])) {
            this.rpcMethod = 'eth_getBlockTransactionCountByHash';
          }
          _get(_getPrototypeOf(GetBlockTransactionCountMethod.prototype), "beforeExecution", this).call(this, moduleInstance);
        }
      }]);
      return GetBlockTransactionCountMethod;
    }(web3CoreMethod.AbstractGetBlockTransactionCountMethod);

    function _createSuper$3(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$3()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$3() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var GetBlockUncleCountMethod = function (_AbstractGetBlockUncl) {
      _inherits(GetBlockUncleCountMethod, _AbstractGetBlockUncl);
      var _super = _createSuper$3(GetBlockUncleCountMethod);
      function GetBlockUncleCountMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, GetBlockUncleCountMethod);
        return _super.call(this, 'eth_getUncleCountByBlockNumber', utils, formatters, moduleInstance);
      }
      _createClass(GetBlockUncleCountMethod, [{
        key: "beforeExecution",
        value: function beforeExecution(moduleInstance) {
          if (this.isHash(this.parameters[0])) {
            this.rpcMethod = 'eth_getUncleCountByBlockHash';
          }
          _get(_getPrototypeOf(GetBlockUncleCountMethod.prototype), "beforeExecution", this).call(this, moduleInstance);
        }
      }]);
      return GetBlockUncleCountMethod;
    }(web3CoreMethod.AbstractGetBlockUncleCountMethod);

    function _createSuper$4(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$4()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$4() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var GetTransactionFromBlockMethod = function (_AbstractGetTransacti) {
      _inherits(GetTransactionFromBlockMethod, _AbstractGetTransacti);
      var _super = _createSuper$4(GetTransactionFromBlockMethod);
      function GetTransactionFromBlockMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, GetTransactionFromBlockMethod);
        return _super.call(this, 'eth_getTransactionByBlockNumberAndIndex', utils, formatters, moduleInstance);
      }
      _createClass(GetTransactionFromBlockMethod, [{
        key: "beforeExecution",
        value: function beforeExecution(moduleInstance) {
          if (this.isHash(this.parameters[0])) {
            this.rpcMethod = 'eth_getTransactionByBlockHashAndIndex';
          }
          _get(_getPrototypeOf(GetTransactionFromBlockMethod.prototype), "beforeExecution", this).call(this, moduleInstance);
        }
      }]);
      return GetTransactionFromBlockMethod;
    }(web3CoreMethod.AbstractGetTransactionFromBlockMethod);

    function _createSuper$5(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$5()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$5() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var RskSignTransactionMethod = function (_SignTransactionMetho) {
      _inherits(RskSignTransactionMethod, _SignTransactionMetho);
      var _super = _createSuper$5(RskSignTransactionMethod);
      function RskSignTransactionMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, RskSignTransactionMethod);
        return _super.call(this, utils, formatters, moduleInstance);
      }
      _createClass(RskSignTransactionMethod, [{
        key: "beforeExecution",
        value: function beforeExecution(moduleInstance) {
          this.parameters[0] = this.formatters.inputTransactionFormatter(this.parameters[0], moduleInstance);
        }
      }, {
        key: "execute",
        value: function execute() {
          if (isString(this.parameters[1])) {
            var account = this.moduleInstance.accounts.wallet[this.parameters[1]];
            if (account) {
              return this.moduleInstance.transactionSigner.sign(this.parameters[0], account.privateKey);
            }
          }
          return _get(_getPrototypeOf(RskSignTransactionMethod.prototype), "execute", this).call(this);
        }
      }]);
      return RskSignTransactionMethod;
    }(web3CoreMethod.SignTransactionMethod);

    function _createSuper$6(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$6()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$6() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var RskSignMethod = function (_SignMethod) {
      _inherits(RskSignMethod, _SignMethod);
      var _super = _createSuper$6(RskSignMethod);
      function RskSignMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, RskSignMethod);
        return _super.call(this, utils, formatters, moduleInstance);
      }
      _createClass(RskSignMethod, [{
        key: "execute",
        value: function execute() {
          if (this.moduleInstance.accounts.wallet[this.parameters[1]]) {
            return this.signLocally();
          }
          return _get(_getPrototypeOf(RskSignMethod.prototype), "execute", this).call(this);
        }
      }, {
        key: "signLocally",
        value: function () {
          var _signLocally = _asyncToGenerator( _regeneratorRuntime.mark(function _callee() {
            var signedMessage;
            return _regeneratorRuntime.wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _context.prev = 0;
                    this.beforeExecution(this.moduleInstance);
                    signedMessage = this.moduleInstance.accounts.sign(this.parameters[1], this.moduleInstance.accounts.wallet[this.parameters[0]].privateKey);
                    if (!this.callback) {
                      _context.next = 6;
                      break;
                    }
                    this.callback(false, signedMessage);
                    return _context.abrupt("return");
                  case 6:
                    return _context.abrupt("return", signedMessage);
                  case 9:
                    _context.prev = 9;
                    _context.t0 = _context["catch"](0);
                    if (!this.callback) {
                      _context.next = 14;
                      break;
                    }
                    this.callback(_context.t0, null);
                    return _context.abrupt("return");
                  case 14:
                    throw _context.t0;
                  case 15:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee, this, [[0, 9]]);
          }));
          function signLocally() {
            return _signLocally.apply(this, arguments);
          }
          return signLocally;
        }()
      }]);
      return RskSignMethod;
    }(web3CoreMethod.SignMethod);

    function _createSuper$7(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$7()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$7() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var RskGetAccountsMethod = function (_GetAccountsMethod) {
      _inherits(RskGetAccountsMethod, _GetAccountsMethod);
      var _super = _createSuper$7(RskGetAccountsMethod);
      function RskGetAccountsMethod(utils, formatters, moduleInstance) {
        _classCallCheck(this, RskGetAccountsMethod);
        return _super.call(this, utils, formatters, moduleInstance);
      }
      _createClass(RskGetAccountsMethod, [{
        key: "execute",
        value: function execute() {
          if (this.moduleInstance.accounts.wallet.accountsIndex) {
            var accounts = [];
            for (var i = 0; i < this.moduleInstance.accounts.wallet.accountsIndex; i++) {
              accounts.push(this.moduleInstance.accounts.wallet[i].address);
            }
            return Promise.resolve(accounts);
          }
          return _get(_getPrototypeOf(RskGetAccountsMethod.prototype), "execute", this).call(this);
        }
      }]);
      return RskGetAccountsMethod;
    }(web3CoreMethod.GetAccountsMethod);

    function _createSuper$8(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$8()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$8() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var MethodFactory = function (_AbstractMethodFactor) {
      _inherits(MethodFactory, _AbstractMethodFactor);
      var _super = _createSuper$8(MethodFactory);
      function MethodFactory(utils, formatters) {
        var _this;
        _classCallCheck(this, MethodFactory);
        _this = _super.call(this, utils, formatters);
        _this.methods = {
          getNodeInfo: web3CoreMethod.GetNodeInfoMethod,
          getProtocolVersion: web3CoreMethod.GetProtocolVersionMethod,
          getCoinbase: web3CoreMethod.GetCoinbaseMethod,
          isMining: web3CoreMethod.IsMiningMethod,
          getHashrate: web3CoreMethod.GetHashrateMethod,
          isSyncing: web3CoreMethod.IsSyncingMethod,
          getGasPrice: web3CoreMethod.GetGasPriceMethod,
          getAccounts: RskGetAccountsMethod,
          getBlockNumber: web3CoreMethod.GetBlockNumberMethod,
          getBalance: web3CoreMethod.GetBalanceMethod,
          getStorageAt: web3CoreMethod.GetStorageAtMethod,
          getCode: web3CoreMethod.GetCodeMethod,
          getBlock: GetBlockMethod,
          getUncle: GetUncleMethod,
          getBlockTransactionCount: GetBlockTransactionCountMethod,
          getBlockUncleCount: GetBlockUncleCountMethod,
          getTransaction: web3CoreMethod.GetTransactionMethod,
          getPendingTransactions: web3CoreMethod.GetPendingTransactionsMethod,
          getTransactionFromBlock: GetTransactionFromBlockMethod,
          getTransactionReceipt: web3CoreMethod.GetTransactionReceiptMethod,
          getTransactionCount: web3CoreMethod.GetTransactionCountMethod,
          sendSignedTransaction: web3CoreMethod.SendRawTransactionMethod,
          signTransaction: RskSignTransactionMethod,
          sendTransaction: web3CoreMethod.EthSendTransactionMethod,
          sign: RskSignMethod,
          call: web3CoreMethod.CallMethod,
          estimateGas: web3CoreMethod.EstimateGasMethod,
          submitWork: web3CoreMethod.SubmitWorkMethod,
          getWork: web3CoreMethod.GetWorkMethod,
          getPastLogs: web3CoreMethod.GetPastLogsMethod,
          requestAccounts: web3CoreMethod.RequestAccountsMethod,
          getId: web3CoreMethod.VersionMethod,
          getChainId: web3CoreMethod.ChainIdMethod,
          getProof: web3CoreMethod.GetProofMethod
        };
        return _this;
      }
      return MethodFactory;
    }(web3CoreMethod.AbstractMethodFactory);

    var SubscriptionsFactory = function () {
      function SubscriptionsFactory(utils, formatters) {
        _classCallCheck(this, SubscriptionsFactory);
        this.utils = utils;
        this.formatters = formatters;
      }
      _createClass(SubscriptionsFactory, [{
        key: "getSubscription",
        value: function getSubscription(moduleInstance, type, options) {
          switch (type) {
            case 'logs':
              return new web3CoreSubscriptions.LogSubscription(options, this.utils, this.formatters, moduleInstance, new web3CoreMethod.GetPastLogsMethod(this.utils, this.formatters, moduleInstance));
            case 'newBlockHeaders':
              return new web3CoreSubscriptions.NewHeadsSubscription(this.utils, this.formatters, moduleInstance);
            case 'pendingTransactions':
              return new web3CoreSubscriptions.NewPendingTransactionsSubscription(this.utils, this.formatters, moduleInstance);
            case 'syncing':
              return new web3CoreSubscriptions.SyncingSubscription(this.utils, this.formatters, moduleInstance);
            default:
              throw new Error("Unknown subscription: ".concat(type));
          }
        }
      }]);
      return SubscriptionsFactory;
    }();

    function _createSuper$9(Derived) { return function () { var Super = _getPrototypeOf(Derived), result; if (_isNativeReflectConstruct$9()) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
    function _isNativeReflectConstruct$9() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }
    var Rsk3 = function (_AbstractWeb3Module) {
      _inherits(Rsk3, _AbstractWeb3Module);
      var _super = _createSuper$9(Rsk3);
      function Rsk3(provider, net) {
        var _this;
        var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
        _classCallCheck(this, Rsk3);
        if (_.isString(provider) && (provider.includes('http:') || provider.includes('ws:'))) {
          console.warn('SECURITY WARNING: connection to RSK node is not secure');
        }
        _this = _super.call(this, provider, options, new MethodFactory(Utils, web3CoreHelpers.formatters), net);
        if (!options.transactionSigner || options.transactionSigner.type === 'TransactionSigner') {
          options.transactionSigner = new TransactionSigner(Utils, web3CoreHelpers.formatters);
        }
        var providerResolver = new web3Providers.ProvidersModuleFactory().createProviderResolver();
        var resolvedProvider = providerResolver.resolve(provider, net);
        _this.net = new rsk3Net.Network(resolvedProvider, null, options);
        _this.accounts = new rsk3Account.Accounts(resolvedProvider, null, options);
        _this.personal = new rsk3Personal.Personal(resolvedProvider, null, _this.accounts, options);
        _this.abiCoder = new rsk3Abi.AbiCoder();
        _this.utils = Utils;
        _this.formatters = web3CoreHelpers.formatters;
        _this.subscriptionsFactory = new SubscriptionsFactory(Utils, web3CoreHelpers.formatters);
        _this.contractModuleFactory = new rsk3Contract.ContractModuleFactory(Utils, web3CoreHelpers.formatters, _this.abiCoder, _this.accounts);
        _this.initiatedContracts = [];
        _this._transactionSigner = options.transactionSigner;
        _this.Contract = function (abi, address) {
          var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
          var contract = _this.contractModuleFactory.createContract(_this.currentProvider, _this.accounts, abi, address, {
            defaultAccount: options.from || options.defaultAccount || _this.defaultAccount,
            defaultBlock: options.defaultBlock || _this.defaultBlock,
            defaultGas: options.gas || options.defaultGas || _this.defaultGas,
            defaultGasPrice: options.gasPrice || options.defaultGasPrice || _this.defaultGasPrice,
            transactionBlockTimeout: options.transactionBlockTimeout || _this.transactionBlockTimeout,
            transactionConfirmationBlocks: options.transactionConfirmationBlocks || _this.transactionConfirmationBlocks,
            transactionPollingTimeout: options.transactionPollingTimeout || _this.transactionPollingTimeout,
            transactionSigner: _this.transactionSigner,
            data: options.data
          });
          _this.initiatedContracts.push(contract);
          return contract;
        };
        return _this;
      }
      _createClass(Rsk3, [{
        key: "clearSubscriptions",
        value: function clearSubscriptions() {
          return _get(_getPrototypeOf(Rsk3.prototype), "clearSubscriptions", this).call(this, 'eth_unsubscribe');
        }
      }, {
        key: "subscribe",
        value: function subscribe(type, options, callback) {
          return this.subscriptionsFactory.getSubscription(this, type, options).subscribe(callback);
        }
      }, {
        key: "setProvider",
        value: function setProvider(provider, net) {
          var setContractProviders = this.initiatedContracts.every(function (contract) {
            return contract.setProvider(provider, net);
          });
          return this.net.setProvider(provider, net) && this.personal.setProvider(provider, net) && this.ens.setProvider(provider, net) && _get(_getPrototypeOf(Rsk3.prototype), "setProvider", this).call(this, provider, net) && setContractProviders;
        }
      }, {
        key: "transactionSigner",
        get: function get() {
          return this._transactionSigner;
        }
        ,
        set: function set(transactionSigner) {
          if (transactionSigner.type && transactionSigner.type === 'TransactionSigner') {
            throw new Error('Invalid TransactionSigner given!');
          }
          this._transactionSigner = transactionSigner;
          this.accounts.transactionSigner = transactionSigner;
          this.ens.transactionSigner = transactionSigner;
          this.initiatedContracts.forEach(function (contract) {
            contract.transactionSigner = transactionSigner;
          });
        }
      }, {
        key: "defaultGasPrice",
        set: function set(value) {
          this.initiatedContracts.forEach(function (contract) {
            contract.defaultGasPrice = value;
          });
          this.net.defaultGasPrice = value;
          this.personal.defaultGasPrice = value;
          _set(_getPrototypeOf(Rsk3.prototype), "defaultGasPrice", value, this, true);
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Rsk3.prototype), "defaultGasPrice", this);
        }
      }, {
        key: "defaultGas",
        set: function set(value) {
          this.initiatedContracts.forEach(function (contract) {
            contract.defaultGas = value;
          });
          this.net.defaultGas = value;
          this.personal.defaultGas = value;
          _set(_getPrototypeOf(Rsk3.prototype), "defaultGas", value, this, true);
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Rsk3.prototype), "defaultGas", this);
        }
      }, {
        key: "transactionBlockTimeout",
        set: function set(value) {
          this.initiatedContracts.forEach(function (contract) {
            contract.transactionBlockTimeout = value;
          });
          this.net.transactionBlockTimeout = value;
          this.personal.transactionBlockTimeout = value;
          _set(_getPrototypeOf(Rsk3.prototype), "transactionBlockTimeout", value, this, true);
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Rsk3.prototype), "transactionBlockTimeout", this);
        }
      }, {
        key: "transactionConfirmationBlocks",
        set: function set(value) {
          this.initiatedContracts.forEach(function (contract) {
            contract.transactionConfirmationBlocks = value;
          });
          this.net.transactionConfirmationBlocks = value;
          this.personal.transactionConfirmationBlocks = value;
          _set(_getPrototypeOf(Rsk3.prototype), "transactionConfirmationBlocks", value, this, true);
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Rsk3.prototype), "transactionConfirmationBlocks", this);
        }
      }, {
        key: "transactionPollingTimeout",
        set: function set(value) {
          this.initiatedContracts.forEach(function (contract) {
            contract.transactionPollingTimeout = value;
          });
          this.net.transactionPollingTimeout = value;
          this.personal.transactionPollingTimeout = value;
          _set(_getPrototypeOf(Rsk3.prototype), "transactionPollingTimeout", value, this, true);
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Rsk3.prototype), "transactionPollingTimeout", this);
        }
      }, {
        key: "defaultAccount",
        set: function set(value) {
          this.initiatedContracts.forEach(function (contract) {
            contract.defaultAccount = Rsk3.utils.toChecksumAddress(value);
          });
          this.net.defaultAccount = value;
          this.personal.defaultAccount = value;
          _set(_getPrototypeOf(Rsk3.prototype), "defaultAccount", value, this, true);
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Rsk3.prototype), "defaultAccount", this);
        }
      }, {
        key: "defaultBlock",
        set: function set(value) {
          this.initiatedContracts.forEach(function (contract) {
            contract.defaultBlock = value;
          });
          this.net.defaultBlock = value;
          this.personal.defaultBlock = value;
          _set(_getPrototypeOf(Rsk3.prototype), "defaultBlock", value, this, true);
        }
        ,
        get: function get() {
          return _get(_getPrototypeOf(Rsk3.prototype), "defaultBlock", this);
        }
      }], [{
        key: "givenProvider",
        get: function get() {
          return web3Providers.ProviderDetector.detect();
        }
      }, {
        key: "utils",
        get: function get() {
          return Utils;
        }
      }, {
        key: "Net",
        get: function get() {
          var providerResolver = new web3Providers.ProvidersModuleFactory().createProviderResolver();
          return function (provider, options, net) {
            return new rsk3Net.Network(providerResolver.resolve(provider, net), options);
          };
        }
      }, {
        key: "Personal",
        get: function get() {
          var providerResolver = new web3Providers.ProvidersModuleFactory().createProviderResolver();
          return function (provider, options, net) {
            return new rsk3Personal.Personal(providerResolver.resolve(provider, net), options);
          };
        }
      }]);
      return Rsk3;
    }(web3Core.AbstractWeb3Module);

    return Rsk3;

})));
